common/src/main/java/org/conscrypt/HpkeContextSender.java

```diff
@@ -16,100 +16,119 @@
-            throw new IllegalStateException(
-                    "Error while setting up base sender with the keys provided", e);
-        }
+    /**
+     * Returns an uninitialised HpkeContextSender from a specific {@link Provider}
+     *
+     * @param suite the HPKE suite to use.  @see {@link HpkeSuite} for details.
+     * @param providerName the name of the Provider to use
+     * @return an uninitialised HpkeContextSender for the requested suite
+     * @throws NoSuchAlgorithmException if no implementation could be found
+     * @throws NoSuchProviderException if providerName is null or no such Provider exists
+     */
+    public static HpkeContextSender getInstance(String suite, String providerName)
+        throws NoSuchAlgorithmException, NoSuchProviderException {
+        return new HpkeContextSender(findSpi(suite, providerName));
     }
 
     /**
-     * Returns the enc (encapsulated key) that was created during the initialization/setup phase.
+     * Returns an uninitialised HpkeContextSender from a specific {@link Provider}
      *
-     * @return enc (encapsulated key)
+     * @param suite the HPKE suite to use.  @see {@link HpkeSuite} for details.
+     * @param provider the Provider to use
+     * @return an uninitialised HpkeContextSender for the requested suite
+     * @throws NoSuchAlgorithmException if no implementation could be found
+     * @throws NoSuchProviderException if provider is null
      */
-    public byte[] getEnc() {
-        return enc;
+    public static HpkeContextSender getInstance(String suite, Provider provider)
+        throws NoSuchAlgorithmException, NoSuchProviderException {
+        return new HpkeContextSender(findSpi(suite, provider));
     }
 
     /**
-     * Hybrid Public Key Encryption (HPKE) encryption.
-     * <p>
-     * Note: This API keeps track of its state. It maintains an internal HPKE context. As a result,
-     * to encrypt multiple messages that are expected to be decrypted using the same context, one
-     * must call this method for each of the messages.
+     * Initialises this HpkeContextSender.
      *
-     * @param plaintext message that will be encrypted
-     * @param aad       optional associated data
-     * @return ciphertext
+     * @param mode HPKE mode to use, currently only HpkeContext.MODE_BASE is supported.
+     * @param publicKey public key of the recipient
+     * @param info application-supplied information, may be null or empty
+     * @throws InvalidKeyException if publicKey is null or an unsupported key format
+     * @throws UnsupportedOperationException if mode is not a supported HPKE mode
+     * @throws IllegalStateException if this HpkeContextSender has already been initialised
      */
-    public byte[] seal(byte[] plaintext, byte[] aad) {
-        Preconditions.checkNotNull(plaintext, "plaintext");
-        return NativeCrypto.EVP_HPKE_CTX_seal(ctx, plaintext, aad);
+    public void init(int mode, PublicKey publicKey, byte[] info)
```

Comment:
An issue I see here is with the "mode" parameter. If mode is BASE_MODE, the rest of the parameters are ok. If later, we expand and support AUTH, PSK, or AUTH_PSK, we don't have all the parameters. My suggestion would be to have 4 different "init" methods to initialize the `HpkeContextSender` and you won't need to pass the mode, it can be inferred. Just adding 1 "init" method should be ideal for now as the other modes are not supported yet.

BASE 
pkR, info

AUTH
pkR, info, skS

PSK
pkR, info,  psk, pid

AUTH & PSK
pkR, info, skS, psk, pid

Reply:
And 3 more for testing, which is a lot of inits.... It's maybe the best way though.

Would probably mean adding extra args now to the engineInit* methods so that can stay stable, otherwise if we need to add extra init methods to HpkSpi then we end up with version skew issues.

Reply:
Done!  The fact that there are defaults for PSK args makes things a bit cleaner, and dropping the mode arg makes the usages a lot cleaner.